servicex/minio_adaptor.py

```diff
@@ -116,6 +100,84 @@ def do_copy() -> None:
         return await asyncio.wrap_future(self._download_executor.submit(do_copy))
 
 
+class MinioAdaptorFactory:
+    '''A factor that will return, when asked, the proper minio adaptor to use for a request
+    to get files from ServiceX.
+    '''
+    def __init__(self, c: Optional[ConfigView] = None,
+                 always_return: Optional[MinioAdaptor] = None):
+        '''Create the factor with a possible adaptor to always use
+
+        Args:
+            always_return (Optional[MinioAdaptor], optional): The adaptor to always use. If none,
+            then one will be created by the best other available method. Defaults to None.
+        '''
+        # Get the defaults setup.
+        self._always = always_return
+        self._config_adaptor = None
+        if self._always is None and c is not None:
+            self._config_adaptor = self._from_config(c)
+
+    def from_best(self, transation_info: Optional[Dict[str, str]] = None) -> MinioAdaptor:
+        '''Using the information we have, create the proper Minio Adaptor with the correct
+        endpoint and login information. Order of adaptor generation:
+
+        1. The `always_return` option from the ctor
+        1. Use info from the `translation_info`
+        1. Use the info from the config handed in to the ctor
+
+        Raises:
+            Exception: If we do not have enough information to create an adaptor, this is raised.
+
+        Returns:
+            MinioAdaptor: The adaptor that can be used to extract the data from minio for this
+            request.
+        '''
+        if self._always is not None:
+            logging.getLogger(__name__).debug('Using the pre-defined minio_adaptor')
+            return self._always
+        if transation_info is not None:
+            if 'minio-endpoint' in transation_info \
+                    and 'minio-access-key' in transation_info \
+                    and 'minio-secret-key' in transation_info:
+                logging.getLogger(__name__).debug('Using the request-specific minio_adaptor')
+                return MinioAdaptor(transation_info['minio-endpoint'],
+                                    transation_info['minio-access-key'],
+                                    transation_info['minio-secret-key'])
+        if self._config_adaptor is not None:
+            logging.getLogger(__name__).debug('Using the config-file minio_adaptor')
+            return self._config_adaptor
+        raise Exception("Do not know how to create a Minio Login info")
```

Comment:
Best practice is to define a custom exception class which subclasses `Exception`.

Reply:
Yeah - probably right - this is me being lazy. I'll update!

Reply:
Ok - PR to remove all Exception throws in the main code! Thanks for the suggestion!